File: 0322-coin-change/0322-coin-change.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        if(amount == 0)
        {
            return 0;
        }
        int dp[++amount];
        dp[0] = 0;
        sort(coins.begin(),coins.end());
        for(int i=1;i<amount;i++)
        {
            dp[i] = INT_MAX;
            for(int coin:coins)
            {
                if(i-coin  < 0)
                {
                    break;
                }
                if(dp[i-coin] != INT_MAX)
                {
                    dp[i] = min(dp[i],1+dp[i-coin]);
                }
            }
        }
        return dp[--amount] == INT_MAX ? -1 : dp[amount];

    }
};

File: 0322-coin-change/0322-coin-change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        INF = amount + 1

        dp = [INF] * (amount + 1)
        dp[0] = 0

        for i in range(1, amount + 1):
            for coin in coins:
                if i >= coin:
                    dp[i] = min(dp[i], dp[i - coin] + 1)

        return -1 if dp[amount] == INF else dp[amount]